build-system/compile/compile.js

```diff
@@ -396,19 +398,20 @@ function compile(entryModuleFilenames, outputDir, outputFilename, options) {
           })
           .pipe(nop())
           .on('end', resolve);
-    } else {
-      return gulp.src(srcs, {base: '.'})
-          .pipe(gulpIf(shouldShortenLicense, shortenLicense()))
-          .pipe(sourcemaps.init({loadMaps: true}))
-          .pipe(gulpClosureCompile(compilerOptionsArray))
-          .on('error', err => {
-            handleCompilerError(outputFilename);
-            reject(err);
-          })
-          .pipe(rename(outputFilename))
-          .pipe(sourcemaps.write('.'))
-          .pipe(gulp.dest(outputDir))
-          .on('end', resolve);
     }
+
+    return gulp.src(srcs, {base: '.'})
+        .pipe(gulpIf(shouldShortenLicense, shortenLicense()))
+        .pipe(sourcemaps.init({loadMaps: true}))
+        .pipe(gulpIf(({path}) => !/third_party/.test(path), babel({plugins})))
+        .pipe(gulpClosureCompile(compilerOptionsArray))
+        .on('error', err => {
+          handleCompilerError(outputFilename);
+          reject(err);
+        })
+        .pipe(rename(outputFilename))
+        .pipe(sourcemaps.write('.'))
+        .pipe(gulp.dest(outputDir))
+        .on('end', resolve);
```

Comment:
Can you measure the running time of `gulp dist` before and after this change? The `gulp-if` on line 405 will be evaluated ~O(N^2) times the number of source files in the repo, so we should be sure this doesn't adversely affect build time.

Reply:
This PR is increasing the running time from 8 mins to 32 mins.

Before: https://travis-ci.org/ampproject/amphtml/jobs/532074185#L305
After: https://travis-ci.org/ampproject/amphtml/jobs/532099597#L466

Let's hope there's a different way to do these transforms.

Reply:
@erwinmombay Any ideas?

Reply:
It seems like the issue is that `compile` is called recursively so transformations are executed multiple times for the same file.

Possible options I can think of:

- Memoize the result of `gulpIf(/^(?!.*third_party)/, babel({plugins}))` by filename.
- Execute transformations before recursive compilation into a temporary folder, and read from there before Closure.

Reply:
@alanorozco I'm pretty sure `compile()` is not called recursively. The issue is that `gulp.src()` reads O(N) source files for each of the O(N) invocations of `compile()`.

Can you help me understand the purpose of the `babel` transformations, and where in the compilation they need to be done, and to which source files?

Re: writing files to a temp folder: The use of `gulp` streams gives us the ability to do things like transforms and license shortening without breaking sourcemaps. There's also a significant performance boost to `gulp dist` from not having to write files to disc for intermediate steps. We should try to retain this advantage as far as possible.

Reply:
> I'm pretty sure compile() is not called recursively.

You're correct, I was mis-remembering from looking at it at 2 AM :)

> Can you help me understand the purpose of the babel transformations, where in the compilation workflow they need to be done, and to which source files?

`extensions` and `src`, possibly `3p` as well?

The transformations have different purposes in general but they all attempt to reduce bundle size. We're also intending to turn on HTML template minification transforms after the `prettier` adoption.

> The use of gulp streams gives us the ability to do things like transforms and license shortening without breaking sourcemaps.

Agreed that we should try to keep these properties.

Reply:
Removed `babel` pass from multipass altogether. So no more slowdown.